File: samplestore/src/main/java/com/stripe/samplestore/StoreCart.java
package com.stripe.samplestore;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Currency;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.UUID;

public class StoreCart implements Parcelable {

    @NonNull private final Currency mCurrency;
    @NonNull private final LinkedHashMap<String, StoreLineItem> mStoreLineItems;

    StoreCart(@NonNull Currency currency) {
        mCurrency = currency;
        // LinkedHashMap because we want iteration order to be the same.
        mStoreLineItems = new LinkedHashMap<>();
    }

    public StoreCart(@NonNull String currencyCode) {
        this(Currency.getInstance(currencyCode));
    }

    void addStoreLineItem(@NonNull String description, int quantity, long unitPrice) {
        addStoreLineItem(new StoreLineItem(description, quantity, unitPrice));
    }

    private void addStoreLineItem(@NonNull StoreLineItem storeLineItem) {
        mStoreLineItems.put(UUID.randomUUID().toString(), storeLineItem);
    }

    public boolean removeLineItem(@NonNull String itemId) {
        return mStoreLineItems.remove(itemId) != null;
    }

    @NonNull
    List<StoreLineItem> getLineItems() {
        return new ArrayList<>(mStoreLineItems.values());
    }

    int getSize() {
        return mStoreLineItems.size();
    }

    @NonNull
    Currency getCurrency() {
        return mCurrency;
    }

    long getTotalPrice() {
        long total = 0L;
        for (StoreLineItem item : mStoreLineItems.values()) {
            total += item.getTotalPrice();
        }
        return total;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mCurrency.getCurrencyCode());
        dest.writeInt(mStoreLineItems.size());
        for (String key : mStoreLineItems.keySet()) {
            dest.writeString(key);
            dest.writeParcelable(mStoreLineItems.get(key), 0);
        }
    }

    private StoreCart(@NonNull Parcel in) {
        mCurrency = Currency.getInstance(in.readString());
        int count = in.readInt();
        mStoreLineItems = new LinkedHashMap<>();
        for (int i = 0; i < count; i++) {
            mStoreLineItems.put(in.readString(),
                    in.readParcelable(StoreLineItem.class.getClassLoader()));
        }
    }

    public static final Parcelable.Creator<StoreCart> CREATOR =
            new Parcelable.Creator<StoreCart>() {
                @Override
                public StoreCart createFromParcel(Parcel source) {
                    return new StoreCart(source);
                }

                @Override
                public StoreCart[] newArray(int size) {
                    return new StoreCart[size];
                }
            };
}

File: samplestore/src/main/java/com/stripe/samplestore/StoreCart.kt
package com.stripe.samplestore

import android.os.Parcel
import android.os.Parcelable

import java.util.ArrayList
import java.util.Currency
import java.util.LinkedHashMap
import java.util.UUID

class StoreCart : Parcelable {

    val currency: Currency
    private val storeLineItems: LinkedHashMap<String, StoreLineItem>

    internal val lineItems: List<StoreLineItem>
        get() = ArrayList(storeLineItems.values)

    internal val totalPrice: Long
        get() {
            var total = 0L
            for (item in storeLineItems.values) {
                total += item.totalPrice
            }
            return total
        }

    internal constructor(currency: Currency) {
        this.currency = currency
        // LinkedHashMap because we want iteration order to be the same.
        storeLineItems = LinkedHashMap()
    }

    fun addStoreLineItem(description: String, quantity: Int, unitPrice: Long) {
        addStoreLineItem(StoreLineItem(description, quantity, unitPrice))
    }

    private fun addStoreLineItem(storeLineItem: StoreLineItem) {
        storeLineItems[UUID.randomUUID().toString()] = storeLineItem
    }

    fun removeLineItem(itemId: String): Boolean {
        return storeLineItems.remove(itemId) != null
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(currency.currencyCode)
        dest.writeInt(storeLineItems.size)
        for (key in storeLineItems.keys) {
            dest.writeString(key)
            dest.writeParcelable(storeLineItems[key], 0)
        }
    }

    private constructor(input: Parcel) {
        currency = Currency.getInstance(input.readString())
        val count = input.readInt()
        storeLineItems = LinkedHashMap()
        for (i in 0 until count) {
            val id = input.readString()!!
            val item = input
                .readParcelable<StoreLineItem>(StoreLineItem::class.java.classLoader)!!
            storeLineItems[id] = item
        }
    }

    companion object CREATOR : Parcelable.Creator<StoreCart> {
        override fun createFromParcel(parcel: Parcel): StoreCart {
            return StoreCart(parcel)
        }

        override fun newArray(size: Int): Array<StoreCart?> {
            return arrayOfNulls(size)
        }
    }
}
